beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -1985,14 +2012,29 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         // Note: this will declare a re-org if we skip `SLOTS_PER_HISTORICAL_ROOT` blocks
         // between calls to fork choice without swapping between chains. This seems like an
         // extreme-enough scenario that a warning is fine.
-        let is_reorg = current_head.block_root
-            != new_head
-                .beacon_state
-                .get_block_root(current_head.slot)
-                .map(|root| *root)
-                .unwrap_or_else(|_| Hash256::random());
+        let is_reorg = new_head
+            .beacon_state
+            .get_block_root(current_head.slot)
+            .map_or(true, |root| *root != current_head.block_root);
+
+        let mut reorg_distance = Slot::new(0);
 
         if is_reorg {
+            match self.find_reorg_slot(&new_head.beacon_state, &new_head.beacon_block_root) {
+                Ok(Some(slot)) => reorg_distance = current_head.slot - slot,
+                // This means there is no intersection between the old and new chains in the past
+                // `SLOTS_PER_HISTORICAL_ROOT` slots. And the current chain length is less than
+                // `SLOTS_PER_HISTORICAL_ROOT`. Default to a reorg distance of 0.
```

Comment:
Thinking about the default value for `depth` when we can't/don't detect the full distance. I can think of these options:

- `0` (current)
    - Pros: 0 doesn't *really* make sense as a re-org distance, so it's kind of like a `None`.
    - Cons: It might cause naive downstream apps to think that there was actually no re-org and become faulty.
- `SLOTS_PER_HISTORICAL_ROOT`
    - Pros: it's closer to the real value than `0`.
    - Cons: Also seems like it might cause downstream apps to be faulty.
- `current_head.slot`
    - Pros: it declares that there was a re-org of the full length of the chain so a downstream app would *technically* be safe because it would check all the slots in the chain.
    - Cons: it's probably wrong and it might cause downstream apps to do a whole bunch of unnecessary work.
- `None` (change our Rust definition so this is possible)
    - Pros: is explicit that we don't know what the distance is.
    - Cons: some downstream apps might not expect this.
    
After typing this out, I'm tempted to go with the `None` approach since it will force the LH code-base to explicitly handle the case where we chose not to determine a re-org distance. I could also make an issue on the API specs repo to make this explicit in a comment regarding the `chain_reorg` event, I think we're well within our rights to decide *not* to compute this value. 
    
What are your thoughts on this?

Reply:
Thinking about this more... rather than iterating backwards until a depth of `SLOTS_PER_HISTORICAL_ROOT` in the worst case, shouldn't we be iterating backwards until the finalized head?

And in the extreme scenario where the finalized head is greater than `SLOTS_PER_HISTORICAL_ROOT` slots ago, if a validator is relying entirely on server sent events, could they function properly if we do anything other than return the actual re-org depth? Or is the re-org depth only necessary to inform something like shuffling recalculation (so only matters if it's over a threshold of 2 epochs or something).

I do think if we do anything other than return the actual re-org depth, we should raise this as a API spec repo issue.

Reply:
> Thinking about this more... rather than iterating backwards until a depth of SLOTS_PER_HISTORICAL_ROOT in the worst case, shouldn't we be iterating backwards until the finalized head?

We could try to calculate a re-org floor via finalization, but it's a little tricky to get right (a re-org can also finalize any number of epochs) and if we're still in the `SLOTS_PER_HISTORICAL_ROOT` array then it's cheap enough to search the whole thing. In other words, I'm not sure the optimization would be worth it.

> I do think if we do anything other than return the actual re-org depth, we should raise this as a API spec repo issue.

I agree, I started the conversation here: https://github.com/ethereum/eth2.0-APIs/pull/127

Reply:
@ajsutton makes a good point here that we could use the finalized slot of the previous has as a "floor" for the re-org distance: https://github.com/ethereum/eth2.0-APIs/pull/127#issuecomment-784734706

This falls into the category of "technically wrong but safe".